File: 2022/days/05/main.ts
import { readFileSync } from 'fs'



function parseInstruction(str: string): number[] {
    let regExp = new RegExp(/move (\d+) from (\d+) to (\d+)/)
    const matches = str.match(regExp)
    if (matches) {
        let matchesOfInterest: number[] = matches.slice(1).map(elem => Number(elem))

        return [matchesOfInterest[0], matchesOfInterest[1] - 1, matchesOfInterest[2] - 1]
    }
    throw new Error(`could not parse instruction: ${str}`);
}

function parseStacks(str: string, stacks: string[][]): string[][] {
    let index = 0
    const width = 4

    while (index < str.length) {
        if (str[index] == '[') {
            stacks[Number((index / width) )].unshift(str[index + 1])
        }
        index += width
    }

    return stacks
}

function moveBoxes(instruction: number[], stacks: string[][], atOnce: number = 1): string[][] {

    let [quantity, from, to] = instruction

    let removed = stacks[from].splice(-atOnce, atOnce)
    stacks[to] = stacks[to].concat(removed)

    if (quantity - atOnce > 0) {
        stacks = moveBoxes([quantity - 1, from, to], stacks, atOnce)
    }

    return stacks
}

enum State {
    Initalising,
    ReadingLabels,
    MovingBoxes
}



const file = readFileSync('input.txt', 'utf-8')

let stacksPartOne: string[][]= Array.from({ length: 9 }, () => []);
let stacksPartTwo: string[][]= Array.from({ length: 9 }, () => []);
let state: State = State.Initalising
const didgetOnlyRegex = new RegExp(/^(?: *\d *)+$/)

file.split('\n').forEach(line => {
    line = line.replace("\r","")

    if(state === State.Initalising){
        if(line.match(didgetOnlyRegex) != null){
            state = State.ReadingLabels
            return
        }
        stacksPartOne = parseStacks(line,stacksPartOne)
        stacksPartTwo = parseStacks(line,stacksPartTwo)
    }

    if(state === State.ReadingLabels){
        if(line == ''){
            state = State.MovingBoxes
            return
        }
    }

    if(state === State.MovingBoxes){
        if(line == ''){
            return
        }
        let instructions: number[] = parseInstruction(line)
        stacksPartOne = moveBoxes(instructions,stacksPartOne)
        stacksPartTwo = moveBoxes(instructions,stacksPartTwo,instructions[0])
    }
})


let accPartOne = ''
let accPartTwo = ''
stacksPartOne.forEach(stack => {
    if(stack.length != 0){
        accPartOne += stack.slice(-1)
    }
});

stacksPartTwo.forEach(stack => {
    if(stack.length != 0){
        accPartTwo += stack.slice(-1)
    }
});
console.log(accPartOne,accPartTwo)

File: 2022/days/05/main.py
import re
from enum import Enum, auto

def parseInitialState(line: str, stacks: list[list[str]]):
    # assuming we have single digit labels
    offset = 0
    boxWidth = 4
    while len(line) > offset:
        if line[offset] == '[':
            stacks[int(offset/boxWidth)] = [line[offset+1]] + stacks[int(offset/boxWidth)]
        offset += boxWidth
    return stacks

def parseMove(line: str):
    matches = re.match(r'move (\d+) from (\d+) to (\d+)', line)
    if matches is None:
        return []
    l = list(map(lambda x: int(x),matches.groups()))
    l[1] -= 1
    l[2] -= 1

    return l

def moveBoxes(stacks: list[list[str]],instructions : list[int], amount:int=1):

    boxes = stacks[instructions[1]][-amount:]
    stacks[instructions[1]] = stacks[instructions[1]][:-amount]
    stacks[instructions[2]] += boxes

    if instructions[0] > 1:
        return moveBoxes(stacks,[instructions[0]-1,instructions[1],instructions[2]],amount)

    return stacks
        

if __name__ == '__main__':
    class State(Enum):
        INIT = auto()
        LABLES= auto()
        MOVING= auto()
    
    fileName = 'example.txt'

    stackPtOne: list[list[str]] = [[]]*9
    stackPtTwo: list[list[str]] = [[]]*9


    state = State.INIT

    partOneAcc = ''
    partTwoAcc = ''
    with open(fileName,'r') as file:
        for line in file:
            if state is State.INIT:
                if re.match(r'^(?: *\d *)+$',line) is not None:
                    state = State.LABLES
                    continue
                else:
                    stackPtOne = parseInitialState(line,stackPtOne)
                    stackPtTwo = parseInitialState(line,stackPtTwo)

            
            if state is State.LABLES:
                if re.match(r'^ *$',line) is not None:
                    state = State.MOVING
                    continue
            
            if state is State.MOVING:
                instructions = parseMove(line)

                stackPtOne = moveBoxes(stackPtOne,parseMove(line))

                stackPtTwo = moveBoxes(stackPtTwo,parseMove(line),amount=parseMove(line)[0])

    for stack in stackPtOne:
        if stack != []:
            partOneAcc += stack[-1]

    for stack in stackPtTwo:
        if stack != []:
            partTwoAcc += stack[-1]
    print(partOneAcc,partTwoAcc)